libsolidity/codegen/ArrayUtils.cpp

```diff
@@ -46,256 +46,67 @@ void ArrayUtils::copyArrayToStorage(ArrayType const& _targetType, ArrayType cons
+	bool haveSourceLengthOnStack = fromCalldata && _sourceType.isDynamicallySized();
 
-	// stack: source_ref [source_length] target_ref
-	// store target_ref
 	for (unsigned i = _sourceType.sizeOnStack(); i > 0; --i)
 		m_context << swapInstruction(i);
 	// stack: target_ref source_ref [source_length]
 
-	if (_targetType.isByteArrayOrString())
+	if (sourceBaseType->category() == Type::Category::Array)
 	{
-		// stack: target_ref source_ref [source_length]
-		if (fromCalldata && _sourceType.isDynamicallySized())
-		{
-			// stack: target_ref source_ref source_length
-			m_context << Instruction::SWAP1;
-			// stack: target_ref source_length source_ref
-			m_context << Instruction::DUP3;
-			// stack: target_ref source_length source_ref target_ref
-			m_context.callYulFunction(
-				m_context.utilFunctions().copyByteArrayToStorageFunction(_sourceType, _targetType),
-				3,
-				0
-			);
-		}
-		else
-		{
-			// stack: target_ref source_ref
-			m_context << Instruction::DUP2;
-			// stack: target_ref source_ref target_ref
-			m_context.callYulFunction(
-				m_context.utilFunctions().copyByteArrayToStorageFunction(_sourceType, _targetType),
-				2,
-				0
-			);
-		}
-		// stack: target_ref
-		return;
+		// TODO: This limitation can now be removed since we use Yul utility functions that handle nested arrays correctly.
+		// The old inline assembly implementation couldn't handle nested calldata dynamic arrays, but the Yul functions
+		// support them through recursive calls. We keep this check temporarily for backward compatibility.
+		auto const& sourceBaseArrayType = dynamic_cast<ArrayType const&>(*sourceBaseType);
+		solUnimplementedAssert(
+			_sourceType.location() != DataLocation::CallData ||
+			!_sourceType.isDynamicallyEncoded() ||
+			!sourceBaseArrayType.isDynamicallySized(),
+			"Copying nested calldata dynamic arrays to storage is not implemented in the old code generator."
+		);
 	}
-
-	// retrieve source length
-	if (_sourceType.location() != DataLocation::CallData || !_sourceType.isDynamicallySized())
-		retrieveLength(_sourceType); // otherwise, length is already there
-	if (_sourceType.location() == DataLocation::Memory && _sourceType.isDynamicallySized())
+	else
 	{
-		// increment source pointer to point to data
-		m_context << Instruction::SWAP1 << u256(0x20);
-		m_context << Instruction::ADD << Instruction::SWAP1;
+		// TODO: This limitation can now be removed since we use Yul utility functions that handle non-value types correctly.
+		// The old inline assembly implementation couldn't handle copying arrays of non-value types from memory or calldata to storage,
+		// but the Yul functions support them. We keep this check temporarily for backward compatibility.
+		bool isMemoryOrCalldata = _sourceType.location() == DataLocation::Memory || _sourceType.location() == DataLocation::CallData;
+		solUnimplementedAssert(
+			!(!sourceBaseType->isValueType() && isMemoryOrCalldata),
```

Comment:
Too much negation makes things harder to read.

Also, the name of `isMemoryOrCalldata` is missing any indication as to whether it's the source or target.

```suggestion
			sourceBaseType->isValueType() || !fromMemoryOrCalldata,
```